toolkit/themes/shared/design-system/config.js

```diff
@@ -108,67 +135,146 @@ ${formattedVars}
- * preserves variable references when necessary.
+ * Finds the original value of a token for a given media query and surface.
+ *
+ * @param {object} token - Token object parsed by style-dictionary.
+ * @param {string} prop - Name of the property we're querying for.
+ * @param {string} surface
+ *  The desktop surface we're generating CSS for, either "brand" or "platform".
+ * @returns {string} The original token value based on our parameters.
+ */
+function getOriginalTokenValue(token, prop, surface) {
+  if (surface) {
+    return token.original.value[surface]?.[prop];
+  } else if (prop == "default" && typeof token.original.value != "object") {
+    return token.original.value;
+  }
+  return token.original.value?.[prop];
+}
+
+/**
+ * Updates a token's value to the relevant original value after resolving
+ * variable references.
  *
  * @param {object} token - Token object parsed from JSON by style-dictionary.
- * @param {string} prop
- *  Name of the property used to get the token's new value.
+ * @param {string} originalVal
+ *  Original value of the token for the combination of surface and media query.
  * @param {object} dictionary
  *  Object of transformed tokens and helper fns provided by style-dictionary.
  * @returns {object} Token object with an updated value.
  */
-function transformTokenValue(token, prop, dictionary) {
-  let originalVal = token.original.value[prop];
-  if (dictionary.usesReference(originalVal)) {
-    let refs = dictionary.getReferences(originalVal);
-    return { ...token, value: `var(--${refs[0].name})` };
+function transformTokenValue(token, originalVal, dictionary) {
+  let value = originalVal;
+  if (dictionary.usesReference(value)) {
+    dictionary.getReferences(value).forEach(ref => {
+      value = value.replace(`{${ref.path.join(".")}}`, `var(--${ref.name})`);
+    });
   }
-  return { ...token, value: token[prop] || originalVal };
+  return { ...token, value };
+}
+
+/**
+ * Creates a light-dark transform that works for a given surface. Registers
+ * the transform with style-dictionary and returns the transform's name.
+ *
+ * @param {string} surface
+ *  The desktop surface we're generating CSS for, either "brand", "platform",
+ *  or "shared".
+ * @returns {string} Name of the transform that was registered.
+ */
+const createLightDarkTransform = surface => {
+  let name = `lightDarkTransform/${surface}`;
+  let matcher = getMatcher(surface);
+  let transformer = getTransformer(surface);
+
+  StyleDictionary.registerTransform({
+    type: "value",
+    transitive: true,
+    name,
+    matcher,
+    transformer,
+  });
+
+  return name;
+};
+
+/**
+ * Determines the correct matcher function to use for a given surface.
+ *
+ * @param {string} surface - Desktop surface, either "brand", "platform", or "shared".
+ * @returns {function}
+ *  Matcher function for determining if a token's value needs to undergo
+ *  a light-dark transform.
+ */
+function getMatcher(surface) {
```

Comment:
These function names should maybe be more specific, like `getLightDarkMatcher` and `getLightDarkTransformer`. It looks like they're only used by `createLightDarkTransform`, so another option would be to move them in that function definition. You could even name them `matcher` and `transformer` then for the sweet sweet property shorthand (and tweak them to take `token` as an arg and remove the inner functions)

Reply:
ah nifty, I like it. Another win for JS closures, hooray!